agreement_dynamic/models/agreement_section.py

```diff
@@ -44,16 +85,72 @@ def onchange_copyvalue(self):
             self.copyvalue = "${{object.{}.{} or {}}}".format(
                 self.field_id.name,
                 self.sub_model_object_field_id.name,
-                self.default_value or "''",
+                self._get_proper_default_value(),
             )
 
     # compute the dynamic content for jinja expression
     def _compute_dynamic_content(self):
-        MailTemplates = self.env["mail.template"]
         for this in self:
-            content = MailTemplates._render_template(
-                this.content, "agreement.section", [this.id]
-            )[this.id]
+            content = this._render_template(
+                this.content,
+                this.resource_ref_model_id.model,
+                this.res_id,
+                datas={"header": Header()},
+            )
             this.dynamic_content = content
 
+    def _get_proper_default_value(self):
+        self.ensure_one()
+        value = "''"
+        if self.default_value:
+            if self.field_id.ttype in ("integer", "float"):
+                value = "{}"
+            else:
+                value = "'{}'"
+            value = value.format(self.default_value)
+        return value
 
+    @api.model
+    def _render_template(self, template_txt, model, res_ids, datas=False):
+        """
+        Render input provided by user, for report and preview
+        It is an edited version of mail.template._render_template()
+        """
+        if isinstance(res_ids, int):
+            res_ids = [res_ids]
+        if datas and not isinstance(datas, dict):
+            raise UserError(_("datas argument is not a proper dict"))
+        results = dict.fromkeys(res_ids, u"")
+        # try to load the template
+        try:
+            mako_env = mako_safe_template_env
+            template = mako_env.from_string(tools.ustr(template_txt))
+        except Exception:
+            return False
```

Comment:
We'll still have to look at this, I've seen you added the code but for some reason it doesnt work well.

Reply:
The exception here is risen normally as a user error when we click validate, but indeed we get an internal server errror when we try to preview this right away. For getting rid of this, I will add a boolean field in agreement, called is_validated, it would depend on content, and will hide the preview button if the validate button has not been pressed successfully first

Reply:
Good solution for now - I think you could solve this by making the Preview button a `type="object"` Python action instead of a report action, validating within that action, and then executing the report action by returning it as a dictionary if correctly validated. That way, the double button press is basically done automatically.